### PR TITLE
Introduce Passphrase Cache and other features

### LINKED ISSUE #25: Use secret storage API to remember passphrase
VS Code provides secret storage API for managing sensitive data across workspaces.

See <https://code.visualstudio.com/api/references/vscode-api#SecretStorage> for more information.

We may use this API to remember passphrases for user in a safe way.

### LINKED ISSUE #40: Reload key when git config singingKey has changed
When editing the git.user.singingKey manually, it still uses the old one until reload. Would be nice if it would use the updated one.

### PR DESCRIPTION
**Pull request by AnnAngela**
This PR introduced these features:
* Passphrase Cache via `SecretStorage` from vscode, resolves #25, with:
  1. Unlock key automatically when the associated passphrase has been found;
  2. Commands for listing, deleting or clearing the cache;
  3. Configuration for enabling or disabling this feature, default `false`.
* Show the associated user id in the status bar element, with a configuration to control which style to use;
* Show the associated user id in the passphrase prompt;
* Automatically update the current key and status bar's visibility from git config in `KeyStatusManager#syncStatus`, resolves #40.

This PR introduced these changed:
* Use `vscode.l10n.t` to support i18n;
* Make a `Map`-style class to handle the use of `SecretStorage`;
* Use `async-lock` for `KeyStatusManager#syncStatus` and `KeyStatusManager#updateFolder` to avoid race conditions;
* Move `VscodeOutputLogger` to `logger.ts` to reduce the size of `extension.ts` (maybe other class should be moved too?);
* Style changes, such as changes to make long lines to multiple shorter lines for code readability.

I know that this PR is huge and I should split it, but these changes are highly coupled, sorry for that.

Feel free to ask my any questions. And to be honest, I am not good at English, so maybe the translation in `l10n/bundle.l10n.json` and `package.nls.json` should be rewriting :(

### CODE REVIEW

**Comment by wdhongtw** [comment]
Some commits are irrelevant to others

- `chore: ignore dist folder`
- `refactor: Move Logger to logger.ts`
- `refactor: Use String#trimEnd directly`

But rest of them should be in one commit, as them all do one logical job. :D

BTW, the commit `npm: Bump @types/node from 18.14.2 to 18.14.4` seems not changing anything in `package.json`?

**Comment by wdhongtw** [comment]
Some advices for the UX flow and messages.

- Most `SECURELY-stored`/`SECURELY-cached` term can be removed or replaced with `stored`/`cached`, It's a little aggressive to claim the security again and again. The one in `keyUnlockedWithCachedPassphrase` should be enough.
- `at all` can be remove from `noCachedPassphrase`, as a tone of blaming.
- `passphraseClearanceConfirm` and `passphraseDeletionConfirm` and related confirming logic is not so appropriate, since that 1 the target users are all programmers, not normal user and 2. the action is destructive for the extension, but not destructive for users, (user can always setup the passphrase cache again.) Thanks for the implementation but I thought these should be removed.

**Comment by AnnAngela** [comment]
> BTW, the commit `npm: Bump @types/node from 18.14.2 to 18.14.4` seems not changing anything in `package.json`?

Yes, it's only affected `package-lock.json`.

**Comment by AnnAngela** [comment]
> Some advices for the UX flow and messages.
> 
> * Most `SECURELY-stored`/`SECURELY-cached` term can be removed or replaced with `stored`/`cached`, It's a little aggressive to claim the security again and again. The one in `keyUnlockedWithCachedPassphrase` should be enough.
> * `at all` can be remove from `noCachedPassphrase`, as a tone of blaming.
> * `passphraseClearanceConfirm` and `passphraseDeletionConfirm` and related confirming logic is not so appropriate, since that 1 the target users are all programmers, not normal user and 2. the action is destructive for the extension, but not destructive for users, (user can always setup the passphrase cache again.) Thanks for the implementation but I thought these should be removed.

For the first and second comments, I'm sorry for that, that have been modified.

For the third comment, I think the user might want to be able to delete the cache, for example when the user abandoned their old keys, maybe they want to delete the cached passphrase, even if it's stored in `SecretStorage`.

File: README.md [hunk 1]
```diff
@@ -23,6 +23,16 @@ signing key, together with a cute icon to tell whether the key is unlocked or no
 
 When you click the indicator, you will be prompted for passphrase to unlock the key.
 
+### Passphrase cache
```
Comment (wdhongtw) [inline README.md:26]: > Passphrase Cache
upper case here :D

File: src/extension.ts [hunk 2]
```diff
@@ -22,42 +79,163 @@ export function activate(context: vscode.ExtensionContext) {
...
+            return;
+        }
+        if ((await vscode.window.showInformationMessage<vscode.MessageItem>(
+            vscode.l10n.t("passphraseDeletionConfirm"),
+            { modal: true, detail: vscode.l10n.t('yourSelection', targets.map(({ label, detail }) => `* ${label}${detail ? `\n  ${detail}` : ""}`).join('\n')) },
+            { title: YES },
+            { title: NO, isCloseAffordance: true },
+        ))?.title !== YES) {
+            return;
+        }
```
Comment (wdhongtw) [inline src/extension.ts:148]: > For the third comment, I think the user might want to be able to delete the cache, for example when the user abandoned their old keys, maybe they want to delete the cached passphrase, even if it's stored in SecretStorage.

Agreed. What I means is that we keep `gpgIndicator.deletePassphraseCache` command but remove the double checking of `passphraseDeletionConfirm` dialog. :D

Same for `gpgIndicator.clearPassphraseCache` and `passphraseClearanceConfirm` below.

**Comment by wdhongtw** [comment]
Once this PR ready, I think I will update `Contributors` section in README.

How would you like to be mentioned? Is AnnAngela a good name for you? :D

File: src/extension.ts (hunk 2 above)
Comment (AnnAngela) [reply src/extension.ts]: I agreed for removing `passphraseDeletionConfirm`, but I think the `passphraseClearanceConfirm` should be kept for possible false click in the command list.

**Comment by AnnAngela** [comment]
> Once this PR ready, I think I will update `Contributors` section in README.
> 
> How would you like to be mentioned? Is AnnAngela a good name for you? :D

That's OK, thank you!

**Comment by AnnAngela** [comment]
Oh, there's a question: shall we popup a notification to disclose these changes after user updating the extension?

File: src/extension.ts (hunk 2 above)
Comment (wdhongtw) [reply src/extension.ts]: OK :D

**Comment by wdhongtw** [comment]
> Oh, there's a question: shall we popup a notification to disclose these changes after user updating the extension?

Sounds good, maybe something like

```mermaid
flowchart LR
    A[ ]
    B{cache\nenabled}
    C{feature\nnotified}
    D[RestJob]
    E{MessageBox\n'Would you like ...'}
    F[notified = true\nenabled = true]
    G[notified = true\nenabled = false]
    H[MessageBox\n'You can turn on later by ...']
    A-->|Unlock Command| B
    B -->|False| C
    B -->|True| D
    C -->|False| E
    C -->|True| D
    E -->|Enable| F
    E -->|Don't show again| H
    H --> G
    F --> D
    G --> D
```

`notified` state can be preserved in <https://code.visualstudio.com/api/references/vscode-api#Memento>

**Comment by AnnAngela** [comment]
Oh, you thought further than me. My original idea was popup a notification after user upgradation to tell people new feature available and left a link to changelog or readme.

But your ider is better. I will work on it after this pr merged and commit another pr to avoid mess 😂

**Comment by wdhongtw** [comment]
@AnnAngela thanks for the great work!

I'm going to merge this PR, could you (rebase and) squash these commits into one commit (and force push again)?

I could do this on GitHub web UI, but your commit signature will be losed if I do this through GitHub Web UI. :(

**Comment by AnnAngela** [comment]
@wdhongtw I tried to squash the commits via github desktop, but it seems to break sth I guess?

**Pull request closed without merging**